components/script/dom/htmltablesectionelement.rs

```diff
@@ -71,6 +76,64 @@ impl HTMLTableSectionElementMethods for HTMLTableSectionElement {
     fn Rows(&self) -> Root<HTMLCollection> {
         HTMLCollection::create(&window_from_node(self), NodeCast::from_ref(self), box RowsFilter)
     }
+
+    // https://html.spec.whatwg.org/multipage/#dom-tbody-insertrow
+    fn InsertRow(&self, index: i32) -> Fallible<Root<HTMLElement>> {
+        if index < -1 {
+            return Err(Error::IndexSize);
+        }
+
+        let self_node = NodeCast::from_ref(self);
+        let tr = HTMLTableRowElement::new(String::from("tr"), None, self_node.owner_doc().r());
+
+        if index == -1 {
+            try!(self_node.AppendChild(NodeCast::from_ref(tr.r())));
+            return Ok(HTMLElementCast::from_root(tr));
+        }
+
+        let mut i = 0;
+        for element in self.Rows().elements_iter() {
+            if i == index {
+                try!(self_node.InsertBefore(NodeCast::from_ref(tr.r()),
+                                            Some(NodeCast::from_root(element)).r()));
+                return Ok(HTMLElementCast::from_root(tr));
+            }
+            i += 1;
+        }
+
+        if i == index {
+            try!(self_node.AppendChild(NodeCast::from_ref(tr.r())));
+            return Ok(HTMLElementCast::from_root(tr));
+        }
+
+        return Err(Error::IndexSize)
```

Comment:
The explicit for loop and indexing is a little difficult to understand at first glance... maybe something like this would be more clear?  Or am I going too far down the path of iterator chaining?

```rust
    let element = self.Rows()
                      .elements_iter()
                      .map(|r| Some(r)))
                      .chain(iter::once(None))
                      .nth(index);
    match element {
        None => Err(Error::IndexSize),
        Some(None) => {
            try!(self_node.AppendChild(NodeCast::from_ref(tr.r())));
            Ok(HTMLElementCast::from_root(tr))
        }
        Some(element) => {
            try!(self_node.InsertBefore(NodeCast::from_ref(tr.r()),
                                        Some(NodeCast::from_root(element)).r()));
            Ok(HTMLElementCast::from_root(tr))
        }
    }
```